Better domains - autocompletion, diagnostics, hover

fda-odoo opened the pull request:
The PR contains all changes and fixs about search domains.
- Autocompletion for fields
- comprehension of fields that contains dots
- fix wrong diagnostics of invalid fields
- hover of fields

fda-odoo committed:
[FIX] server: get_full_model_symbols

on models, get_symbols returns all symbols that build the model, but there was no easy way to get all symbols that build the model with inherited models. get_full_model_symbols do it now. It doesn't contain inheritS howevers.

fda-odoo committed:
[IMP] server: use inheritS information for get_full_model_symbols

fda-odoo committed:
[IMP] server: precompute module depencies to improve perfomances

fda-odoo committed:
[IMP] server: add new odoo method signatures

Add return values for with_company, with_context, with_prefetch, with_user, with_env, exists

fda-odoo committed:
[IMP] server: Add expected_type for function arguments

fda-odoo committed:
[IMP] server: autocomplete field, operator and give snippet to build search domain tuple

fda-odoo committed:
[IMP] server: show autocompletion for _inherit models

mmahrouss commented on server/src/features/completion.rs:
```diff
@@ -565,7 +575,48 @@ fn complete_string_literal(session: &mut SessionInfo, file: &Rc<RefCell<Symbol>>
+                let from_module = file.borrow().find_module().clone();
+                Symbol::all_members(&parent, session, &mut all_symbols, true, from_module, &mut None, false);
+                for (_symbol_name, symbols) in all_symbols {
+                    //we could use symbol_name to remove duplicated names, but it would hide functions vs variables
```
But we cannot have functions and variables with the same name, can we?

fda-odoo replied:
We could. Some would be shadowed, but as we are searching here in classes, inheritance, co-models,... we could find different symbols type with the same name

mmahrouss reviewed (approved):
Nice really cool! 
Just one question on a comment, otherwise looks great!

fda-odoo committed:
[IMP] server: improve dot notation diagnostics in search domains

fda-odoo committed:
[IMP] server: improve contexts management

Contexts was not really consistent and wrongly used.
This rework will state that contexts passed as parameters to functions like follow_ref are global contexts linked to the evaluation of the current expression.
EvaluationSymbolWeak now contains a Context, that is a local Context to this symbol, able to hold specific informations that would not be in the tree.
It is mostly used for hooks and odoo code. For example, it will store used __init__ parameter in my_field = fields.Many2one("a_model"), so we can know the linked a_model
to that field later on.
This commit also include some rework on EvaluationSymbolWeak and Ptr and most functions now returns Ptr instead of Weak, to be able to express different kind of values.

fda-odoo committed:
[FIX] server: avoid loop function validation + fix super boolean in attribute evaluation

fda-odoo committed:
[FIX] server: display Any as type in hover if variable has no evaluation

mmahrouss committed:
[FIX] server: fix passed context vals

mmahrouss committed:
[FIX] server: indicate super on hover

fda-odoo committed:
[FIX] server: validation reload conditions for functions in a call expr

fda-odoo committed:
[IMP] server: diagnostics for dot notation in search domains

fda-odoo committed:
[IMP] server: autocompletion for dot notation in domain search
